### ISSUE TITLE
SCFFT_Allocator not ABI compatible

### ISSUE THREAD

**Issue by Spacechild1** [supercollider/supercollider#4438]
The Supercollider plugin interface uses a class `SCFFT_Allocator` in the interface methods `fSCfftCreate` and `fSCfftDestroy`.

Unfortunately, `SCFFT_Allocator` is an abstract virtual base class and calling virtual methods across DLL boundaries should be avoided (unless you use COM techniques*). Different compilers can have different vtable layouts and your plugin may fail/crash in mysterious ways because the wrong virtual method is getting called. Practical example: recent GCC uses two (!) virtual methods for the virtual destructor while MSVC uses one, this means that subsequent virtual methods will be offset by 1.

The allocator should rather be a C struct like:

```
typedef void *(*AllocFunc)(void * user, size_t size);
typedef void (*AsyncStageFn)(void * user, void *ptr);

struct SCFFT_Allocator {
    void *mUser;
    AllocFunc mAlloc;
    FreeFunc mFree;
};

class SCWorld_Allocator : public SCFFT_Allocator {
public:
    SCWorld_Allocator(InterfaceTable* ft, World* world)
        // casting function pointers and calling them without casting back to the original type
        // is technically UB, but in this case it works because World * and void * have the same size
        : mUser((void *)world), mAlloc((AllocFunc)(ft->fRTAlloc)), mFree((FreeFunc)(ft->fRTFree)) {}
};
```

This could be done for the next time the plugin API version changes. UGens using `SCWorld_Allocator` just have to recompiled, but if they use their own subclass of `SCFFT_Allocator `(unlikely), they must change some code.

**Comment by dyfer** [comment]
I came across a misterious [issue with PartConv](https://github.com/supercollider/supercollider/issues/4289) (it uses FFT). This goes beyond my understanding of C++ but I was wondering whether anyone has idea whether this could be related?

**Comment by Spacechild1** [comment]
I don't think this is related. This issue only affects plugins compiled with a different compiler than scsynth/supernova, so it shouldn't affect core plugins.

**Comment by dyfer** [comment]
Thank you for clarifying!

**Comment by sonoro1234** [comment]
How is that we dont have crashes because of this?

**Comment by Spacechild1** [comment]
Note that SC3 plugins are always shipped together with each release and are therefore compiled with the same compiler. But if you build SC3 plugins with MinGW and try to use them with a MSVC Supercollider - or vice versa -, FFT plugins will certainly crash. (I've stumbled upon the above mentioned vtable layout issues while designing a C++ plugin interface for one of my apps.) 

Anyway, my main point is that using C++ classes with virtual methods across plugin boundaries kind of defeats the whole purpose of having a C plugin interface.

**Comment by sonoro1234** [comment]
But also because of having different runtime libraries, it is not good to mix modules compiled with different compilers.

**Comment by Spacechild1** [comment]
> it is not good to mix modules compiled with different compilers.

But that's the whole point of a plugin interface! Take VST plugins for example: some of my VST plugins are from the year 2003 (!) and they still run on all kind of apps (both MSVC and MinGW). Also, most 10+ year old Pd externals still run on latest Pd vanilla. 

With the current situation of Supercollider + SC3 plugins, ABI compatibility is not a big issue, but once the project moves to an open plugin eco-system, this becomes very important!

**Comment by sonoro1234** [comment]
VST and PD avoid linking with runtime libraries?

**Comment by Spacechild1** [comment]
You can't avoid linking to the runtime library but if you link dynamically you have to make sure to ship/install them (on Windows, several versions of MSVCR/MSVCP can coexist). Or you can just link statically (many VST plugins seem to do this, I've checked a couple of them with Dependency Walker).

When compiling with MinGW, you should at least link statically with `-static-libgcc`*) and probably also `-static-libstdc++`. Maybe also `-static -lpthread` for multithreading if your host program doesn't already use libwinpthreads. These are the compiler/linker flags usually used for Pd externals compiled with MinGW: https://github.com/pure-data/pd-lib-builder/blob/1137cca2fadffc1fbb8f0086165ab1076d5100ef/Makefile.pdlibbuilder#L638

Anyway, in Pd all kinds of C/C++ plugins compiled with either VS or MinGW coexist peacefully because the plugin interface is a pure C interface.

*) Under the hood, MinGW always dynamically links to the private MSVCRT.dll, an old version of the MSVC runtime library which is always present, but not really intended to be used by user programs. MinGW does this to avoid shipping any Windows libraries (for licensing reasons, if I remember correctly).

**Comment by sonoro1234** [comment]
Yes, there are workarounds that seem to work in PD or VST but there exist also a lot of issues raised here and there about this: https://stackoverflow.com/questions/26304531/compiling-with-static-libgcc-static-libstdc-still-results-in-dynamic-depende is just one.

I have had alot of issues with this in the past so that finally, my personal decision was not to try this again in the future.

Of course your proposal wouldn't do any harm to people compiling all with the same compiler, so even my skepticism about this, I wouldn't oppose to your proposal.

**Comment by Spacechild1** [comment]
The link you've posted is about Linux, though. There the situation is different because plugins are either available via the package manager (and therefore guaranteed to be compatible) or have to be compiled from source. Pure C plugins tend to work across distro versions but C++ plugins sometimes create problems because of conflicting libstdc++ versions (on Linux, `-static-libstdc++` is rather frowned upon). I'm not really an expert, but I do know that dynamic symbol lookup works differently on Linux and Windows. See for example the RTLD_DEEPBIND flag for dlopen (https://stackoverflow.com/questions/34073051/when-we-are-supposed-to-use-rtld-deepbind): This ensures that a plugin starts to look for symbols within itself and its own dependencies - which AFAICT is the default behavior on Windows. *)

But you're right that there can be indeed issues with static linking under MinGW, but most of the time it just works. Note, however, that ABI compatibility is also about being able to use plugins compiled with an older version of the same compiler. Currently, FFT UGens outside of SC3 plugins are a bit of a gamble (probably not if everything has been compiled with MSVC because Windows has had a stable vtable layout for ages because of COM).

BTW, there are other parts of the SC plugin interface which are problematic, e.g. the audio/control bus locks for Supernova. While the NRT lock is properly exposed via API functions, Supernova's audio/control bus locks are accessed directly, which can create additional problems if both sides have been compiled with different boost/nova versions...

*) I've run into this issues with [vstplugin~] for Pd when trying to load Camomile (Pd as a VST plugin). The issue was that Camomile is statically linked with its own custom Pd version and needs to get the symbols from *that* Pd and not the one which hosts the plugin. This worked out of the box on Windows and macOS but on Linux I had to dlopen the Camomile plugin with RTLD_DEEPBIND.

**Comment by Spacechild1** [comment]
Long story short: multithreaded C++ plugins built with MSVC can continue to work over decades if the plugin interface is a pure C interface (as VST plugins demonstrate). MinGW *mostly* works fine too but there are some additional caveats.

**Comment by Spacechild1** [comment]
Closed by https://github.com/supercollider/supercollider/pull/7329
